### ISSUE TITLE
Sync with cascade delete rules

### ISSUE THREAD

**Issue by DJ-Glock** [paulw11/Seam3#40]
Hello Paul.
Can you please advise if Seam3 supports cascade delete rule? I have set cascade removal for my tables, works with no issues with CoreData itself, but sync fails with unwrapping optional value in some cases.
<img width="1137" alt="2017-09-22 10 38 16" src="https://user-images.githubusercontent.com/26511191/30734152-b3dc39d0-9f82-11e7-9d0c-cba15e61d773.png">

Code is pretty simple:
```
    class func removeTable (table: TablesTable) {
        context.delete(table)
        try? context.save()
    }
```
I have cascade deletion rule from Table to TableSession, and also from TableSession to 2 other entities. If it's needed to remove only one referenced object, for example from GuestsTable or OrdersTable - it works. If it's needed to remove 2 objects, one from GuestsTable, the second - from OrdersTable, it fails.
(For some reason GitHub cannot upload this screenshot.
https://imgur.com/a/dvXg7

**Comment by paulw11** [comment]
When you get the crash can you enter `po results.last!` in the debugger?   I would like to see what object I have if it isn't an `NSManagedObject`

**Comment by DJ-Glock** [comment]
Here you are. I hope I did everything properly, have not used po before.
```
fatal error: unexpectedly found nil while unwrapping an Optional value
(lldb) po results.last!
fatal error: unexpectedly found nil while unwrapping an Optional value
error: Execution was interrupted, reason: EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0).
The process has been returned to the state before expression evaluation.
(lldb) 
```
BTW I can see that array results is blank, contains no values:
<img width="1141" alt="2017-09-23 1 20 10" src="https://user-images.githubusercontent.com/26511191/30766400-7b559a36-9ffd-11e7-9622-be57f80d759d.png">

**Comment by paulw11** [comment]
Can you try with the code from the repo and see if it has fixed your issue?

**Comment by DJ-Glock** [comment]
Well, now it fails on the next step, when I'm trying to save context:
```
    class func removeTable (table: TablesTable) {
        context.delete(table)
        try? context.save()
    }
```
<img width="1136" alt="2017-09-23 1 55 44" src="https://user-images.githubusercontent.com/26511191/30767073-6055c53a-a002-11e7-8b71-30d5f7df9e98.png">
And nothing is synced.

**Comment by paulw11** [comment]
What is your delete rule on the `tableSession` relationship in the `TablesTable` entity?  It should probably be `nullify` since I presume that deleting a single table shouldn't remove the entire `TableSessionTable` entity as *that* may still refer to other `TablesTable` entities.

**Comment by DJ-Glock** [comment]
Nope, I have set it to cascade. Deleting Tables object should delete all TableSessions objects, referenced to this Table. This is rare case for users I believe, but valid. I want to keep the database clean.
<img width="662" alt="2017-09-23 19 19 20" src="https://user-images.githubusercontent.com/26511191/30775010-2449db02-a094-11e7-9e81-57c66949a4ed.png">
And Nullify from TableSessions. If TableSession object is removed, link to it should be just removed from Tables object.
<img width="665" alt="2017-09-23 12 27 43" src="https://user-images.githubusercontent.com/26511191/30771924-a609e07c-a05a-11e7-8d0d-4a74e434fbdc.png">

As per [Apple documentation](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/CoreData/HowManagedObjectsarerelated.html), it should work as expected - when I remove object from Tables entity, it will remove all TableSession objects for this table. But not all TableSessions objects.

> Cascade
> Delete the objects at the destination of the relationship when you delete the source.
> 
> For example, if you delete a department, fire all the employees in that department at the same time.

**Comment by DJ-Glock** [comment]
Hello Paul
What can you suggest here? Will it be fixed by you or should I perform workaround? I believe I can set delete rule to Nullify to all tables or to some of them and handle removal referenced objects manually in code.

**Comment by paulw11** [comment]
Hi.  I will look at this, but there have been a flurry of issues lately and I have my day job too :).
If you can put together a simple test app that reproduces the issue and put it up as a gist or repo on GitHub or even a zip file on dropbox etc then that will help me a lot; I can then concentrate on fixing the issue rather than having to reproduce it first.

**Comment by DJ-Glock** [comment]
Understood :-)

Well, actually I don't have much time to create sample app too. But I can zip my workspace with project and send it to you with all required repro steps. (or can make it public here). Tell me if it will be okay for you, I'll try to do it tomorrow. We have 7 hours difference, so it will be T+1 for you.

**Comment by paulw11** [comment]
You can send me a zip file

**Comment by DJ-Glock** [comment]
Hello Paul.
Please find below repro steps, applicable for my app.

**Device**: any
**iOS**: 10+ (have some issues with iOS 9 for now, so don't rely on it).
**Internet/iCloud accessibility**: not required.

**Repro steps**:
1. Run app
2. Open Tables
<img width="331" alt="2017-09-27 18 41 36" src="https://user-images.githubusercontent.com/26511191/30923487-a3666d88-a3b4-11e7-990f-eba6db5850ef.png">

3. Tap + to add table, add some table
<img width="285" alt="2017-09-27 18 41 53" src="https://user-images.githubusercontent.com/26511191/30923553-d460544e-a3b4-11e7-88e9-94aaf271e545.png">

4. Open created table.
5. Tap on Guests - Plus button 1-2 times to add 2 or more guests.
6. As a result objects in TableSessionsTable, GuestsTable entities will be created.

<img width="337" alt="2017-09-27 18 42 13" src="https://user-images.githubusercontent.com/26511191/30923648-1dc4b10c-a3b5-11e7-8a06-fe51301ee956.png">
7. Hit back button and swipe to left.
8. Remove table
<img width="336" alt="2017-09-27 18 42 28" src="https://user-images.githubusercontent.com/26511191/30923675-31e7421c-a3b5-11e7-9e5c-fcfae55878d1.png">

9. App will crash. Happy debugging! :-)

I have uploaded zip with my project and workspace. Hope you will not face any issues with it.
URL: https://www.dropbox.com/s/on755w5fxba946w/CafeManager.zip?dl=0
Password: Seam3IsCool

BTW. I will really appreciate if you can take a look at my code and project as experienced developer and tell me how my code looks like. Is it good to read and understand?

**Comment by DJ-Glock** [comment]
Hi Paul.
I have one off topic question about conflict resolution.
I have not understood from Seam3 code, it's still bit complicated for me. So asking here.

Is it guaranteed that if I am getting conflict during sync, both conflicting records belong to the same Entity? I don't understand how recordID is being generated.

**Comment by paulw11** [comment]
Yes, the conflicts are detected by the Cloud Kit process, so it hasn't even got to Core Data yet.  During the sync operation, Cloud Kit determines that its local copy of a record and the server copy of a record are in conflict.  

Seam3 simply applies the nominated conflict resolution policy to tell Cloud Kit what the resolved record should look like; either the server version, the local version or the version returned by the conflict resolution closure if that is the option selected. 

You will notice that the records passed to conflict resolution block are `CKRecord`s, not `NSManagedObjects.
